**CLEANING Rates_of_Cases_or_Deaths.csv** (First dataset)

In [1]:
import pandas as pd
import shutil
import re

In [2]:
raw = pd.read_csv("raw/Rates_of_Cases_or_Deaths.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'raw/Rates_of_Cases_or_Deaths.csv'

In [3]:
print(raw)

NameError: name 'raw' is not defined

MISSING AND INVALID DATA

In [8]:
# number of missing values(data)
raw.isnull()

outcome  month  MMWR week  Age group  Vaccine product  \
0       False  False      False      False            False   
1       False  False      False      False            False   
2       False  False      False      False            False   
3       False  False      False      False            False   
4       False  False      False      False            False   
...       ...    ...        ...        ...              ...   
1586    False  False      False      False            False   
1587    False  False      False      False            False   
1588    False  False      False      False            False   
1589    False  False      False      False            False   
1590    False  False      False      False            False   

      Vaccinated with outcome  Fully vaccinated population  \
0                       False                        False   
1                       False                        False   
2                       False                        False   
3                       False                        False   
4                       False                        False   
...                       ...                          ...   
1586                    False                        False   
1587                    False                        False   
1588                    False                        False   
1589                    False                        False   
1590                    False                        False   

      Unvaccinated with outcome  Unvaccinated population  Crude vax IR  \
0                         False                    False         False   
1                         False                    False         False   
2                         False                    False         False   
3                         False                    False         False   
4                         False                    False         False   
...                         ...                      ...           ...   
1586                      False                    False         False   
1587                      False                    False         False   
1588                      False                    False         False   
1589                      False                    False         False   
1590                      False                    False         False   

      Crude unvax IR  Crude IRR  Age adjusted vax IR  Age adjusted unvax IR  \
0              False      False                 True                   True   
1              False      False                 True                   True   
2              False      False                 True                   True   
3              False      False                 True                   True   
4              False      False                 True                   True   
...              ...        ...                  ...                    ...   
1586           False      False                 True                   True   
1587           False      False                False                  False   
1588           False      False                False                  False   
1589           False      False                False                  False   
1590           False      False                False                  False   

      Age adjusted IRR  Continuity correction  
0                 True                  False  
1                 True                  False  
2                 True                  False  
3                 True                  False  
4                 True                  False  
...                ...                    ...  
1586              True                  False  
1587             False                  False  
1588             False                  False  
1589             False                  False  
1590             False                  False  

[1591 rows x 16 columns]

In [14]:
raw.isnull().sum()

outcome                          0
month                            0
MMWR week                        0
Age group                        0
Vaccine product                  0
Vaccinated with outcome          0
Fully vaccinated population      0
Unvaccinated with outcome        0
Unvaccinated population          0
Crude vax IR                     0
Crude unvax IR                   0
Crude IRR                      114
Age adjusted vax IR            987
Age adjusted unvax IR          987
Age adjusted IRR               987
Continuity correction            0
dtype: int64

In [15]:
# % of missing values in each column: number of missing values/total values * 100
raw.isnull().sum()/len(raw)*100

outcome                         0.000000
month                           0.000000
MMWR week                       0.000000
Age group                       0.000000
Vaccine product                 0.000000
Vaccinated with outcome         0.000000
Fully vaccinated population     0.000000
Unvaccinated with outcome       0.000000
Unvaccinated population         0.000000
Crude vax IR                    0.000000
Crude unvax IR                  0.000000
Crude IRR                       7.165305
Age adjusted vax IR            62.036455
Age adjusted unvax IR          62.036455
Age adjusted IRR               62.036455
Continuity correction           0.000000
dtype: float64

In [10]:
# Check for invalid entries in the outcome column
valid_values = ['case', 'death']
invalid_entries = raw[~raw['outcome'].isin(valid_values)]

# Display rows with invalid entries (if any)
print("Invalid entries:")
print(invalid_entries)

# Remove rows with invalid entries
clean = raw[raw['outcome'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['outcome'].unique())

Invalid entries:
Empty DataFrame
Columns: [outcome, month, MMWR week, Age group, Vaccine product, Vaccinated with outcome, Fully vaccinated population, Unvaccinated with outcome, Unvaccinated population, Crude vax IR, Crude unvax IR, Crude IRR, Age adjusted vax IR, Age adjusted unvax IR, Age adjusted IRR, Continuity correction]
Index: []
Unique values after cleaning: ['case' 'death']


In [11]:
# Check for invalid entries in the month column.
# Define a valid pattern for month-year
def is_valid_month(value):
    return bool(re.match(r"^(Jan|Feb|Mar|Apr|May|June|July|Aug|Sept|Oct|Nov|Dec)(2021|2022)$", value))

# Identify invalid entries
invalid_months = raw[~raw['month'].apply(is_valid_month)]

# Display invalid entries
print("Invalid entries in 'month' column:")
print(invalid_months)

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_months)}")

# Remove rows with invalid entries
clean[clean['month'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['month'].unique())

Invalid entries in 'month' column:
     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674       

In [12]:
clean

outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2                        1847                 7.428405e+06   
3                        1558                 7.487444e+06   
4                        1672                 1.439810e+07   
...                       ...                          ...   
1586                      328                 7.078966e+06   
1587                      603                 1.428525e+08   
1588                       33                 1.075017e+07   
1589                      286                 4.941886e+07   
1590                      280                 8.263597e+07   

      Unvaccinated with outcome  Unvaccinated population  Crude vax IR  \
0                         30785              17556462.00     21.687858   
1                         76736              31091322.00     26.503160   
2                         98436              41074337.00     24.864019   
3                         50324              22948569.00     20.808169   
4                         14880               7675907.00     11.612644   
...                         ...                      ...           ...   
1586                        172                963095.40      4.633445   
1587                        398              50965568.85      0.422114   
1588                        398              50965568.85      0.306972   
1589                        398              50965568.85      0.578726   
1590                        398              50965568.85      0.338836   

      Crude unvax IR  Crude IRR  Age adjusted vax IR  Age adjusted unvax IR  \
0         175.348541   8.085102                  NaN                    NaN   
1         246.808418   9.312415                  NaN                    NaN   
2         239.653290   9.638558                  NaN                    NaN   
3         219.290362  10.538667                  NaN                    NaN   
4         193.853313  16.693297                  NaN                    NaN   
...              ...        ...                  ...                    ...   
1586       17.859082   3.854385                  NaN                    NaN   
1587        0.780919   1.850021             0.287258               1.605239   
1588        0.780919   2.543944             0.317448               1.605239   
1589        0.780919   1.349376             0.294841               1.605239   
1590        0.780919   2.304715             0.277278               1.605239   

      Age adjusted IRR  Continuity correction  
0                  NaN                      0  
1                  NaN                      0  
2                  NaN                      0  
3                  NaN                      0  
4                  NaN                      0  
...                ...                    ...  
1586               NaN                      1  
1587          5.588152                      1  
1588          5.056694                      1  
1589          5.444415                      1  
1590          5.789287                      1  

[1591 rows x 16 columns]

In [ ]:
# Save clean to csv file
clean.to_csv("clean_data_1.csv", index=False)

In [16]:
clean = pd.read_csv("clean/clean_data_1.csv")
print(clean)

     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2        

In [17]:
# checking for invalid age groups
unique_age_groups = clean['Age group'].unique()
print(unique_age_groups)

['12-17' '18-29' '30-49' '50-64' '65-79' '80+' 'all_ages_adj' '5-11']


In [18]:
# valid age group intervals 
expected_age_groups = unique_age_groups = [
    "12-17", "18-29", "30-49", "50-64", "65-79", "80+", "5-11", "all_ages_adj"
]

# Check if all values in the 'Age group' column are valid
unexpected_age_groups = clean[~clean['Age group'].isin(expected_age_groups)]

print("Unexpected Age group values:")
print(unexpected_age_groups['Age group'].unique())

# Number of unexpected age groups
print(f"Number of unexpected age groups: {len(unexpected_age_groups)}")

Unexpected Age group values:
[]
Number of unexpected age groups: 0


In [19]:
## Age adjusted vax IR, Age adjusted unvax IR, and Age adjusted IRR have a high % of missing values. 
## As they are important for the analysis, we will replace them by age group metrics.
clean = raw.drop(["Age adjusted vax IR", "Age adjusted unvax IR", "Age adjusted IRR"], axis=1)
print(clean)

     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2        

In [20]:
clean.head()

outcome     month  MMWR week Age group Vaccine product  \
0    case  APR 2021     202114     12-17       all_types   
1    case  APR 2021     202114     18-29       all_types   
2    case  APR 2021     202114     30-49       all_types   
3    case  APR 2021     202114     50-64       all_types   
4    case  APR 2021     202114     65-79       all_types   

   Vaccinated with outcome  Fully vaccinated population  \
0                        8                      36887.0   
1                      674                    2543093.0   
2                     1847                    7428405.0   
3                     1558                    7487444.0   
4                     1672                   14398099.0   

   Unvaccinated with outcome  Unvaccinated population  Crude vax IR  \
0                      30785               17556462.0     21.687858   
1                      76736               31091322.0     26.503160   
2                      98436               41074337.0     24.864019   
3                      50324               22948569.0     20.808169   
4                      14880                7675907.0     11.612644   

   Crude unvax IR  Crude IRR  Continuity correction  
0      175.348541   8.085102                      0  
1      246.808418   9.312415                      0  
2      239.653290   9.638558                      0  
3      219.290362  10.538667                      0  
4      193.853313  16.693297                      0

In [27]:
## Filling "Crude IRR" as % of missing values is 7.1653:
## Calculate mean of this column and fill the missing values with it.

In [21]:
clean = pd.read_csv("clean/clean_data_1.csv")

In [24]:
# calculate the mean of the Crude IRR column
mean_value = clean["Crude IRR"].mean()
print(mean_value)

7.014692613974949


In [23]:
## fillna
clean.fillna({"Crude IRR": mean_value}, inplace=True)

In [18]:
clean.isnull().sum()

outcome                        0
month                          0
MMWR week                      0
Age group                      0
Vaccine product                0
Vaccinated with outcome        0
Fully vaccinated population    0
Unvaccinated with outcome      0
Unvaccinated population        0
Crude vax IR                   0
Crude unvax IR                 0
Crude IRR                      0
Continuity correction          0
dtype: int64

In [19]:
clean.columns

Index(['outcome', 'month', 'MMWR week', 'Age group', 'Vaccine product',
       'Vaccinated with outcome', 'Fully vaccinated population',
       'Unvaccinated with outcome', 'Unvaccinated population', 'Crude vax IR',
       'Crude unvax IR', 'Crude IRR', 'Continuity correction'],
      dtype='object')

DUPLICATES

In [85]:
clean.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1586    False
1587    False
1588    False
1589    False
1590    False
Length: 1591, dtype: bool

In [86]:
clean.duplicated().sum()

np.int64(0)

In [87]:
duplicate_count = int(clean.duplicated().sum())
duplicate_count

0

DATETIME

In [25]:
clean["month"] = pd.to_datetime(clean["month"])

C:\Users\weare\AppData\Local\Temp\ipykernel_3372\2522053795.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  clean["month"] = pd.to_datetime(clean["month"])


In [26]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1591 entries, 0 to 1590
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   outcome                      1591 non-null   object        
 1   month                        1591 non-null   datetime64[ns]
 2   MMWR week                    1591 non-null   int64         
 3   Age group                    1591 non-null   object        
 4   Vaccine product              1591 non-null   object        
 5   Vaccinated with outcome      1591 non-null   int64         
 6   Fully vaccinated population  1591 non-null   float64       
 7   Unvaccinated with outcome    1591 non-null   int64         
 8   Unvaccinated population      1591 non-null   float64       
 9   Crude vax IR                 1591 non-null   float64       
 10  Crude unvax IR               1591 non-null   float64       
 11  Crude IRR                    1591 non-null 

In [107]:
case_rows = clean[clean["outcome"] == "case"].shape[0]
print("Number of case rows:", case_rows)

Number of case rows: 812


In [110]:
death_rows = clean[clean["outcome"] == "death"].shape[0]
print("Number of death rows:", death_rows)

Number of death rows: 779


In [ ]:
source = "clean_data_1.csv"
destination = "clean/"
shutil.move(source, destination)

**CLEANING County_Level_of_community_Transmission.csv** (second dataset)

In [27]:
raw = pd.read_csv("raw/County_Level_of_community_Transmission.csv")

In [28]:
print(raw)

           state_name        county_name  fips_code        date  \
0               Texas        Hays County      48209  01/27/2020   
1               Idaho         Gem County      16045  01/27/2020   
2            Nebraska        Boyd County      31015  10/15/2020   
3               Texas        Knox County      48275  11/03/2020   
4               Idaho   Nez Perce County      16069  11/03/2020   
...               ...                ...        ...         ...   
3225217     Louisiana  Assumption Parish      22007  10/18/2022   
3225218  South Dakota        Hyde County      46069  10/18/2022   
3225219         Texas      Loving County      48301  10/18/2022   
3225220      Virginia     Lynchburg City      51680  10/18/2022   
3225221       Wyoming    Converse County      56009  10/18/2022   

        cases_per_100K_7_day_count_change  \
0                                   0.000   
1                                   0.000   
2                              suppressed   
3              

In [69]:
raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3225222 entries, 0 to 3225221
Data columns (total 7 columns):
 #   Column                                              Dtype  
---  ------                                              -----  
 0   state_name                                          object 
 1   county_name                                         object 
 2   fips_code                                           int64  
 3   date                                                object 
 4   cases_per_100K_7_day_count_change                   object 
 5   percent_test_results_reported_positive_last_7_days  float64
 6   community_transmission_level                        object 
dtypes: float64(1), int64(1), object(5)
memory usage: 172.2+ MB


In [91]:
# fips_code does not seem important comlumn, drop fips_code
clean = raw.drop(["fips_code"], axis=1)
clean.to_csv("clean_data_2.csv", index=False)

In [92]:
print(clean)

           state_name        county_name        date  \
0               Texas        Hays County  01/27/2020   
1               Idaho         Gem County  01/27/2020   
2            Nebraska        Boyd County  10/15/2020   
3               Texas        Knox County  11/03/2020   
4               Idaho   Nez Perce County  11/03/2020   
...               ...                ...         ...   
3225217     Louisiana  Assumption Parish  10/18/2022   
3225218  South Dakota        Hyde County  10/18/2022   
3225219         Texas      Loving County  10/18/2022   
3225220      Virginia     Lynchburg City  10/18/2022   
3225221       Wyoming    Converse County  10/18/2022   

        cases_per_100K_7_day_count_change  \
0                                   0.000   
1                                   0.000   
2                              suppressed   
3                                 354.803   
4                                 544.447   
...                                   ...   
3225217     

In [73]:
clean.describe()

percent_test_results_reported_positive_last_7_days
count                                       2.823884e+06 
mean                                        1.154418e+01 
std                                         1.076036e+01 
min                                         0.000000e+00 
25%                                         3.790000e+00 
50%                                         8.640000e+00 
75%                                         1.618000e+01 
max                                         1.000000e+02

In [74]:
# clean.describe() does not calculate metrics in column cases_per_100K_7_day_count_change 
# because there are non-numerical entries in this column.
# so, convert non-numeric to numeric.
clean['cases_per_100K_7_day_count_change'] = pd.to_numeric(clean['cases_per_100K_7_day_count_change'], 
errors = 'coerce')

In [75]:
clean.describe()

cases_per_100K_7_day_count_change  \
count                       2.468979e+06   
mean                        1.898672e+02   
std                         2.016008e+02   
min                        -9.018800e+02   
25%                         3.879500e+01   
50%                         1.277070e+02   
75%                         2.730585e+02   
max                         9.999970e+02   

       percent_test_results_reported_positive_last_7_days  
count                                       2.823884e+06   
mean                                        1.154418e+01   
std                                         1.076036e+01   
min                                         0.000000e+00   
25%                                         3.790000e+00   
50%                                         8.640000e+00   
75%                                         1.618000e+01   
max                                         1.000000e+02

Interpretation

Cases_per_100K_7_day_count_change

count: 2,468,979 valid non-null values. 
mean: The average 7-day case count change per 100,000 people is 189.87. 
std: The standard deviation is 201.60, indicating significant variability in the case count changes. 
min: The lowest change is -901.88 cases, which could indicate a large decrease in cases during a 7-day period. 
Percentiles: 25%: 38.80 (25% of the data is below this value). 50% (median): 127.71 (half the data is below this value). 
             75%: 273.06 (25% of the data is above this value). max: 999.997 cases per 100,000, likely indicating a very high increase.

Percent_test_results_reported_positive_last_7_days

count: 2,823,884 valid non-NaN values. 
mean: The average positivity rate over the last 7 days is 11.54%. 
std: The standard deviation is 10.76%, indicating some variation in positivity rates. 
min: The lowest positivity rate is 0%, indicating areas with no positive cases in the last 7 days. 
Percentiles: 25%: 3.79%. 50% (median): 8.64%. 75%: 16.18%. max: 100%, suggesting some areas are reporting exclusively positive results.

A positivity rate of 100% may indicate areas where testing is limited or only symptomatic individuals are being tested. 
The wide range from 0% to 100% suggests a significant disparity in how COVID-19 impacts different regions.

In [81]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3225222 entries, 0 to 3225221
Data columns (total 6 columns):
 #   Column                                              Dtype         
---  ------                                              -----         
 0   state_name                                          object        
 1   county_name                                         object        
 2   date                                                datetime64[ns]
 3   cases_per_100K_7_day_count_change                   float64       
 4   percent_test_results_reported_positive_last_7_days  float64       
 5   community_transmission_level                        object        
dtypes: datetime64[ns](1), float64(2), object(3)
memory usage: 147.6+ MB


In [30]:
clean['cases_per_100K_7_day_count_change'] = pd.to_numeric(
    clean['cases_per_100K_7_day_count_change'], errors='coerce'
).fillna(0)
print(clean[cases_per_100K_7_day_count_change])

KeyError: 'cases_per_100K_7_day_count_change'

In [31]:
if 'cases_per_100K_7_day_count_change' in clean.columns:
    print(clean['cases_per_100K_7_day_count_change'])
else:
    print("Column 'cases_per_100K_7_day_count_change' not found in DataFrame.")

Column 'cases_per_100K_7_day_count_change' not found in DataFrame.


CONVERTING DATE

In [80]:
# convert date column into a datetime object.
clean['date'] = pd.to_datetime(clean['date'], format='%m/%d/%Y', errors='coerce')

MISSING VALUES

In [82]:
clean.isnull()

state_name  county_name   date  cases_per_100K_7_day_count_change  \
0             False        False  False                              False   
1             False        False  False                              False   
2             False        False  False                              False   
3             False        False  False                              False   
4             False        False  False                              False   
...             ...          ...    ...                                ...   
3225217       False        False  False                              False   
3225218       False        False  False                              False   
3225219       False        False  False                              False   
3225220       False        False  False                              False   
3225221       False        False  False                              False   

         percent_test_results_reported_positive_last_7_days  \
0                                                     True    
1                                                     True    
2                                                    False    
3                                                     True    
4                                                    False    
...                                                    ...    
3225217                                              False    
3225218                                               True    
3225219                                              False    
3225220                                              False    
3225221                                              False    

         community_transmission_level  
0                                True  
1                                True  
2                               False  
3                               False  
4                               False  
...                               ...  
3225217                         False  
3225218                         False  
3225219                         False  
3225220                         False  
3225221                         False  

[3225222 rows x 6 columns]

In [89]:
clean.isnull().sum()

state_name                                              2662
county_name                                             2662
date                                                       0
cases_per_100K_7_day_count_change                          0
percent_test_results_reported_positive_last_7_days    401338
community_transmission_level                           22661
dtype: int64

In [85]:
# % of missing values in each column = number of missing values/total values * 100
clean.isnull().sum()/len(clean) * 100

state_name                                             0.082537
county_name                                            0.082537
date                                                   0.000000
cases_per_100K_7_day_count_change                      0.000000
percent_test_results_reported_positive_last_7_days    12.443733
community_transmission_level                           0.702618
dtype: float64

In [93]:
# Because the % of missing values is low in state_name column, 
# We'll fillin with a placeholder 'unknown'instead of mode.
clean.fillna({"state_name": "unknown"}, inplace=True)

In [94]:
# Do the same for county_name column. 
clean.fillna({"county_name": "unknown"}, inplace=True)

In [103]:
clean.isnull().sum()

state_name                                            0
county_name                                           0
date                                                  0
cases_per_100K_7_day_count_change                     0
percent_test_results_reported_positive_last_7_days    0
community_transmission_level                          0
dtype: int64

In [97]:
# Filling column 'percent_test_results_reported_positive_last_7_days'.
# This column has numeric values. We'll fillin with a median_value.
median_value = clean["percent_test_results_reported_positive_last_7_days"].median()
print(median_value)
clean.fillna({"percent_test_results_reported_positive_last_7_days": 8.64}, inplace=True)

8.64


In [102]:
# Filling column 'community_transmission_level' with a mode.
# we use iloc[0] just in case there are more than one mode.
mode_value = clean["community_transmission_level"].mode().iloc[0]
print(mode_value)
clean.fillna({"community_transmission_level": "mode_value"}, inplace=True)

high


CHECKING DUPLICATES

In [104]:
clean.duplicated().sum()
duplicat_count = int(clean.duplicated().sum())
duplicat_count

2658

In [105]:
duplicated_rows = clean[clean.duplicated()]
print(duplicated_rows)

       state_name county_name        date cases_per_100K_7_day_count_change  \
148       unknown     unknown  10/17/2022                             0.000   
260       unknown     unknown  10/17/2022                             0.000   
307       unknown     unknown  10/17/2022                             0.000   
327       unknown     unknown  10/17/2022                             0.000   
391       unknown     unknown  10/17/2022                             0.000   
...           ...         ...         ...                               ...   
790316    unknown     unknown  10/17/2022                             0.000   
790317    unknown     unknown  10/17/2022                             0.000   
790318    unknown     unknown  10/17/2022                             0.000   
790319    unknown     unknown  10/17/2022                             0.000   
790320    unknown     unknown  10/17/2022                             0.000   

        percent_test_results_reported_positive_last

In [ ]:
# Ignore duplicate because these rows are not really duplicated. They are filled with the same values(int, or object).

In [107]:
source = "clean_data_2.csv"
destination = "clean/"
shutil.move(source, destination)

'clean/clean_data_2.csv'